### ISSUE TITLE
counters may be per-core, but spec assumes they're per-hart

### ISSUE THREAD

**Issue by timsifive** [riscv/riscv-debug-spec#388]
From: Grigorios Magklis <grigorios.magklis@esperantotech.com>

I have a comment regarding dcsr.stopcount. The spec says:

0: Increment counters as usual.
1: Don't increment any counters while in Debug Mode or on ebreak instructions that cause entry into Debug Mode. These counters include the cycle and instret CSRs. This is preferred for most debugging scenarios.
An implementation may choose not to support writing to this bit. The debugger must read back the value it writes to check whether the feature is supported.

 An interesting situation arises on SMT cores, where multiple harts execute on the same core. The RISC-V spec says:

[...] the cycle CSR which holds a count of the number of clock cycles executed by the processor core on which the hart is running from an arbitrary start time in the past.

In this case, it is legal to implement a single cycle counter which would be shared among all the harts of the processor core (that would make sense to do to save power/area/etc). Interestingly, the Arm architecture also defines the cycle counter (PMCCNTR_EL0) as per-core, not per-hart (hart="PE" in the Arm architecture) [see Arm ARM "D5.1.1 Time as measured by the Performance Monitors cycle counter"]. A quick check at the Intel manual (Chapter 18) also specifies a per-core cycle counter and in newer microprocessors there is a 2nd per-thread cycle counter). By the way, the Intel architecture specifies that the cycle counter counts "unhalted" cycles only; but my reading is that on an SMT processor you would have to halt all threads for the per-core counter to stop (see Table 18-2).

The way the debug spec is written, a legal, multithreaded implementation of RISC-V that doesn't -- unnecessarily -- implement per-hart cycle counters, would have to tie dcsr.stopcount to 1 to be compliant with the debug spec, although it could stop all other counters except the cycle counter.

In short, I think that it would be useful if we could distinguish the cycle counter from the other counters. I propose a small change to the wording in the spec to say that the cycle counter may continue counting even when dcsr.stopcount is 1 while there are harts still running on the core (or we could relax the spec even more and have dcsr.stopcount actually stopping the cycle counter an implementation option). I think that would not be a problem, as existing debug software already knows how to deal with the per-core cycle counters.

**Comment by timsifive** [comment]
The solution I think is to state that this bit only applies to hart-local counters. I'm waiting on feedback on https://github.com/riscv/riscv-isa-manual/issues/249 so I can explain in detail what the intent is.
